pkg/syncer/util.go

```diff
@@ -196,3 +202,16 @@ func migratedToAnnotationExists(ctx context.Context, annotations map[string]stri
 	log.Debugf("%v annotation not found", common.AnnMigratedTo)
 	return false
 }
+
+// GetSCNameFromPVC gets name of the storage class from provided PVC
+func GetSCNameFromPVC(pvc *v1.PersistentVolumeClaim) (string, error) {
+	scName := pvc.Spec.StorageClassName
+	if scName == nil || *scName == "" {
+		scNameFromAnnotation := pvc.Annotations[scNameAnnotationKey]
```

Comment:
For Vanilla Kubernetes, we can not retrieve SC from PVC annotation. 
Is this logic custom logic used in the PSP?

Reply:
This diff resolves the issue [PR 2613381](https://bugzilla.eng.vmware.com/show_bug.cgi?id=2613381). The PR was raised after discussion between Cormac Hogan and Subbu (see section 11 of https://confluence.eng.vmware.com/pages/viewpage.action?pageId=698028013). Based on the discussion, I think Subbu come to the conclusion that we should support reading storage class name through annotation. This diff does not update or add the said annotation in any way while creating/ updating pvc.

Reply:
@subramanian-neelakantan @SandeepPissay @xing-yang 

I feel we should not be using `volume.beta.kubernetes.io/storage-class` annoation which is already deprecated by Kubernetes in 1.6 release.

@bhavyachoudhary Can you remove references to internal vmware links. On Github we should never use internal vmware links.

Reply:
Even though the storage-class annotation is deprecated, the support has not been removed from k8s. There are even tests [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/testing-manifests/statefulset/nginx/statefulset.yaml#L29) that are active to test this support inside a StatefulSet. Until this is completely removed from k8s, the vSphere CSI plugin should continue support too. Otherwise there might be users with old yamls that still specify the storage-class annotation and will be broken with vSphere Volume placement.

Reply:
I agree with @divyenpatel.  I don't think we should add new support for a feature that is already deprecated.  It could be removed any time.

Reply:
I am confused. @xing-yang and @divyenpatel what is the new support we are adding here? Can you kindly make your stance clear?

If you are saying that we are adding support for `storage-class` annotation by this change, that's not true. This is an existing feature in k8s Volumes.

Instead if you are saying that we don't need to support `storage-class` annotation in our feature since that is deprecated, then I disagree. Something that is only "deprecated" and not yet actually "removed" could still be used by any user expecting that to work. So I am not sure what is your stance here.

Reply:
@subramanian-neelakantan @xing-yang I look into this today and tried out creating new volume using SC specified in the PVC annotation. It still works.

```apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-with-sc-annotation
  annotations:
     volume.beta.kubernetes.io/storage-class: example-vanilla-block-sc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
```

```
# kubectl describe  pvc pvc-with-sc-annotation
Name:          pvc-with-sc-annotation
Namespace:     default
StorageClass:  example-vanilla-block-sc
Status:        Bound
Volume:        pvc-e4283b85-c705-4f9a-b5c0-3017165f70b8
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-class: example-vanilla-block-sc
               volume.beta.kubernetes.io/storage-provisioner: csi.vsphere.vmware.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      5Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    <none>
Events:
  Type    Reason                 Age                    From                                                                                                 Message
  ----    ------                 ----                   ----                                                                                                 -------
  Normal  Provisioning           5m28s                  csi.vsphere.vmware.com_vsphere-csi-controller-64cb9bfdb7-6s7lt_352abf83-5a13-4732-8844-35da05e10163  External provisioner is provisioning volume for claim "default/pvc-with-sc-annotation"
  Normal  ExternalProvisioning   5m24s (x2 over 5m28s)  persistentvolume-controller                                                                          waiting for a volume to be created, either by external provisioner "csi.vsphere.vmware.com" or manually created by system administrator
  Normal  ProvisioningSucceeded  5m16s                  csi.vsphere.vmware.com_vsphere-csi-controller-64cb9bfdb7-6s7lt_352abf83-5a13-4732-8844-35da05e10163  Successfully provisioned volume pvc-e4283b85-c705-4f9a-b5c0-3017165f70b8
```

If users are still allowed to create new volumes using SC specified in the PVC annotation. we can add this support.